### PR TITLE
Fail an unpromotable replica shard if its refresh fails

### PR DESCRIPTION
**Pull request by arteam**
In case a refresh for an unpromotable shard fails, we need to fail the search shard(s).

That's an alternative implementation of #94433 where we fail actual replicas unpromotable shards in `TransportBroadcastUnpromotableAction` instead of the local indexing shard in `PostWriteRefresh`.

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java [hunk 1]
```diff
@@ -82,6 +90,30 @@ protected void doExecute(Task task, Request request, ActionListener<ActionRespon
...
+            shardRouting.shardId(),
+            shardRouting.allocationId().getId(),
+            clusterState.metadata().index(shardRouting.getIndexName()).primaryTerm(shardRouting.shardId().getId()),
+            true,
+            "mark unpromotable copy as stale after refresh failure",
+            e,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(Void unused) {
+                    // TODO Only fail if were unable to mark the shard as stale?
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java:104]: Yes. If we successfully fail the shard, no need to fully fail the broadcast request. We might add an debug log message though.

**Review by Tim-Brooks — COMMENTED** [review]
I'll take a closer look tomorrow. But I did note something.

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java [hunk 2]
```diff
@@ -23,8 +23,12 @@ public class UnpromotableShardRefreshRequest extends BroadcastUnpromotableReques
...
     private final long segmentGeneration;
 
-    public UnpromotableShardRefreshRequest(IndexShardRoutingTable indexShardRoutingTable, long segmentGeneration) {
-        super(indexShardRoutingTable);
+    public UnpromotableShardRefreshRequest(
+        IndexShardRoutingTable indexShardRoutingTable,
+        long segmentGeneration,
+        boolean failShardOnError
+    ) {
+        super(indexShardRoutingTable, failShardOnError);
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java:31]: Why do we need this flag? Shouldn't we always fail the search shard if its refresh fails?

File: server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java [hunk 3]
```diff
@@ -82,6 +90,30 @@ protected void doExecute(Task task, Request request, ActionListener<ActionRespon
         }
     }
 
+    private void failShard(ShardRouting shardRouting, ClusterState clusterState, ActionListener<Object> l, Exception e) {
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java:93]: Can we parameterize this with the specific type? `ActionListener<TransportResponse.Empty> l`

I think if you change the call site a bit it should be possible:

```
                request.indexShardRoutingTable.unpromotableShards().forEach(shardRouting -> {
                    final DiscoveryNode node = clusterState.nodes().get(shardRouting.currentNodeId());
                    ActionListener<TransportResponse.Empty> acquired = listeners.acquire(ignored -> {});
                    transportService.sendRequest(
                        node,
                        transportUnpromotableAction,
                        request,
                        TransportRequestOptions.EMPTY,
                        new ActionListenerResponseHandler<>(
                            request.failShardOnError()
                                ? acquired.delegateResponse((l, e) -> failShard(shardRouting, clusterState, l, e))
                                : acquired,
                            (in) -> TransportResponse.Empty.INSTANCE,
                            executor
                        )
                    );
                });
```

**Review by Tim-Brooks — COMMENTED** [review]
Looks pretty good. One comment.

File: server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java (hunk 2 above)
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java:31]: The refresh api propagates an error and does not fail the shard.

**Comment by arteam** [comment]
Thanks Tim!

**Pull request merged**